docs/gsl-intro.md

```diff
@@ -0,0 +1,327 @@
+
+vector<int> v(200);
+process_ints(v); // deduces correct length: 200 (constructs the span from a container)
+~~~
+
+and conveniently supports modern C++ argument initialization when the calling code does have distinct pointer and length arguments:
+
+~~~cpp
+auto remainder = find(v.begin(), v.end(), some_value);
+    // now call process_ints() to fill the rest of the container from *remainder to the end
+process_ints({remainder, v.end()}); // correct and clear (constructs the span from an iterator pair)
+~~~
+
+> Things to remember
+> - Prefer `span` instead of (pointer, length) pairs.
+> - Pass a `span` like a pointer (i.e., by value for "in" parameters). Treat it like a pointer range.
+
+
+## span and const: "What's the difference between `span<const T>` and `const span<T>`?"
+
+`span<const T>` means that the `T` objects are read-only. Prefer this by default, especially as a parameter, if you don't need to modify the `T`s.
+
+`const span<T>` means that the `span` itself can't be made to point at a different target.
+
+`const span<const T>` means both.
+
+> Things to remember
+> - Prefer a `span<const T>` by default to denote that the contents are read-only, unless you do need read-write access.
+
+
+## Iteration: "How do I iterate over a span?"
+
+A `span` is an encapsulated range, and so can be visited using a range-based `for` loop.
+
+Consider the implementation of a function like the `process_ints` that we saw in an earlier example. Visiting every object using a (pointer, length) pair requires an explicit index:
+
+~~~cpp
+void dangerous_process_ints(int* p, size_t n) {
+    for (auto i = 0; i < n; ++i) {
+        p[i] = next_character();
+    }
+}
+~~~
+
+A `span` supports range-`for` -- note this is zero-overhead and does not need to perform any range check, because the range-`for` loop is is known by construction not to exceed the range's bounds:
+
+~~~cpp
+void process_ints(span<int> s) {
+    for (auto& c : s) {
+        c = next_character();
+    }
+}
+~~~
+
+A `span` also supports normal iteration using `.begin()` and `.end()`.
+
+Note that you cannot compare iterators from different spans, even if they refer to the same array.
+
+An iterator is valid as long as the `span` that it is iterating over exists.
+
+
+## Element access: "How do I access a single element in a span?"
+
+Use `myspan[offset]` to subscript, or equivalently use `iter + offset` wheren `iter` is a `span<T>::iterator`. Both are range-checked.
+
+
+
+## Sub-spans: "What if I need a subrange of a span?"
+
+To refer to a sub-span, use `first`, `last`, or `subspan`.
+
+~~~cpp
+void process_ints(span<widget> s) {
+    if (s.length() > 10) {
+        read_header(s.first(10));   // first 10 entries
+        read_rest(s.subspan(10));   // remaining entries
+        // ...
+    }
+}
+~~~
+
+In rarer cases, when you know the number of elements at compile time and want to enable `constexpr` use of `span`, you can pass the length of the sub-span as a template argument:
+
```

Comment:
@jwakely: Are you aware that there are two types of spans? One that keeps the size as a member variable and one that encodes it as part of its type (via its `Extent` template parameter). This API can produce such spans with a fixed size (given that the original type has fixed size too). I'm not aware of any way to achieve this with constexpr member functions only. 

 those API's potentially create a different object type than the noremal `subspan` function? The "normal" `subspan` will always create a dynamically sized span, but the templated ones can r
 (`span` can have the size encoded as part of its type instead of as member variable)